File: src/channels/slack.ts
/**
 * Slack Channel Adapter (OpenClaw Compatible)
 */

import { BaseChannelAdapter } from './base.js';
import type {
  ChannelCapability,
  OutboundMessage,
  PluginContext,
} from '../runtime/types.js';

export interface SlackConfig {
  botToken: string;
  signingSecret: string;
  appToken?: string;
  defaultChannel?: string;
}

export class SlackChannel extends BaseChannelAdapter {
  readonly name = 'slack';
  readonly capabilities: ChannelCapability[] = [
    'text',
    'markdown',
    'images',
    'files',
    'typing',
    'reactions',
    'reply',
    'threads',
  ];

  private config?: SlackConfig;
  private apiBase = 'https://slack.com/api';

  async onInit(ctx: PluginContext): Promise<void> {
    await super.onInit(ctx);
    this.config = ctx.getConfig<SlackConfig>();
  }

  async onStart(): Promise<void> {
    await super.onStart();
    if (!this.config?.botToken) {
      throw new Error('[slack] botToken is required');
    }
    // Verify token
    await this.apiCall('auth.test');
    this.ctx?.logger.info('[slack] Bot connected successfully');
  }

  async send(message: OutboundMessage): Promise<void> {
    this.validateMessage(message);
    const content = await this.formatForChannel(message.content);
    
    const payload: Record<string, unknown> = {
      channel: message.to,
      text: content,
    };

    if (message.options?.threadId) {
      payload.thread_ts = message.options.threadId;
    }

    await this.apiCall('chat.postMessage', payload);
  }

  async formatForChannel(content: string): Promise<string> {
    // Slack uses mrkdwn format
    return content
      .replace(/\*\*([^*]+)\*\*/g, '*$1*') // Bold
      .replace(/\*([^*]+)\*/g, '_$1_') // Italic
      .replace(/`([^`]+)`/g, '`$1`') // Code
      .replace(/```(\w+)?\n?([\s\S]*?)```/g, '```$1\n$2```'); // Code blocks
  }

  async handleUpdate(update: unknown): Promise<void> {
    const event = (update as { event?: unknown }).event;
    if (event) {
      this.ctx?.events.emit('channel:message', { channel: this.name, event });
    }
  }

  private async apiCall(method: string, payload: Record<string, unknown> = {}): Promise<unknown> {
    const response = await fetch(`${this.apiBase}/${method}`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${this.config?.botToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });
    const data = await response.json() as { ok: boolean; error?: string };
    if (!data.ok) throw new Error(`Slack API error: ${data.error}`);
    return data;
  }
}

File: plugins/channels/slack.go
// Slack 通道实现
package channels

import (
	"context"
	"fmt"
)

// SlackConfig Slack 配置
type SlackConfig struct {
	BotToken   string `json:"bot_token"`
	AppToken    string `json:"app_token"`
	SigningSecret string `json:"signing_secret"`
}

// SlackChannel Slack 通道
type SlackChannel struct {
	config *SlackConfig
}

// NewSlackChannel 创建 Slack 通道
func NewSlackChannel(config *SlackConfig) *SlackChannel {
	return &SlackChannel{config: config}
}

// Start 启动通道
func (c *SlackChannel) Start(ctx context.Context) error {
	fmt.Println("Starting Slack channel...")
	return nil
}

// Stop 停止通道
func (c *SlackChannel) Stop(ctx context.Context) error {
	fmt.Println("Stopping Slack channel...")
	return nil
}

// Send 发送消息
func (c *SlackChannel) Send(ctx context.Context, msg *SlackMessage) (string, error) {
	return "msg_ts", nil
}

// ChannelType 返回通道类型
func (c *SlackChannel) ChannelType() string {
	return "slack"
}

// SlackMessage Slack 消息
type SlackMessage struct {
	Channel     string            `json:"channel"`
	Text        string            `json:"text"`
	ThreadTS    string            `json:"thread_ts,omitempty"`
	Attachments []SlackAttachment `json:"attachments,omitempty"`
}

// SlackAttachment Slack 附件
type SlackAttachment struct {
	Color  string `json:"color,omitempty"`
	Title  string `json:"title,omitempty"`
	Text   string `json:"text,omitempty"`
	Fields []struct {
		Title string `json:"title"`
		Value string `json:"value"`
		Short bool   `json:"short"`
	} `json:"fields,omitempty"`
}
